### ISSUE TITLE
Calling HTTPClient.connect in new thread blocks the main thread if there are some problems with connection to DNS

### ISSUE THREAD

**Issue by lukaskotik** [godotengine/godot#7881]
**Operating system or device - Godot version:**
Linux 64 bit, Android, Godot 2.1.2

**Issue description:**

Calling `HTTPClient.connect` in a new thread during an outage of internet service provider blocked main thread for about 20 s. Then it produces the `getaddrinfo failed!` If connected, everything works as expected. If I disconnect ethernet connection, it produces this error immediately without blocking main thread for 20 s.

**Steps to reproduce:**
In connection settings of internet connection, set some weird DNS server, e.g. `5.232.219.115`
Then run the following script. After 1 s, when calling `http.connect("google.com", 80)` blocked the main thread for about 20 s (hence no `delta` is printed).
```
extends Node

var thread = Thread.new()
var thread_runnig = false
var time = 0.0

func _ready():
	set_process(true)

func _process(delta):
	time += delta
	if time > 1.0 and !thread_runnig:
		thread_runnig = true
		print("Starting new thread")
		thread.start(self, "thread_fun", 0)
	print(delta)
		
func thread_fun(par = 0):
	var http = HTTPClient.new()
	print(OS.get_ticks_msec())
	var err = http.connect("google.com", 80) 
	print(err)
	print(OS.get_ticks_msec())
```

**Link to minimal example project:**
:

**Comment by Faless** [comment]
I can confirm this.
It's actually even worse since even the `IP` resolve queue has the same problem (which I think is the suggested way of doing for async resolving).
```
extends Node

var done = false
var time = 0.0

func _ready():
	set_process(true)

func _process(delta):
	time += delta
	if not done:
		print(time)
		if time >= 2.0:
			print("start resolving")
			queue_resolve()
			print("done resolving")

func queue_resolve(par = 0):
	done = true
	return IP.resolve_hostname_queue_item("omfglol.com")
```

The culprit seems to be the heavy usage of `GLOBAL_LOCK_FUNCTION` in the various functions in `core/io/ip.cpp`.
**I think it's only used to avoid race conditions with the internal resolver thread when accessing the cache and queue, so it could be replaced by a local mutex.**
Can @reduz or @punto- comment on this solution? Those locks where there since the initial open source release and I don't want to introduce side effects (threads are evil). But I can work on a PR if you think it might be a proper solution.
I have confirmed that by not calling the `GLOBAL_LOCK_FUNCTION` the bug disappears (but of course strange things might happen without using a mutex for the resolver's cache and queue)

**Comment by marcelofg55** [comment]
@lukaskotik @Faless I've been trying to fix this issue, can you test the attached patch and let me know if it works okay? I've tested it myself and seems working fine but I'd like some more testing, thanks!

[ip.patch.zip](https://github.com/godotengine/godot/files/979408/ip.patch.zip)

**Comment by bojidar-bg** [comment]
@marcelofg55 Are you sure you don't want to make a PR instead? (so that it is simpler to view the patch + comment on specific lines)

**Comment by Faless** [comment]
@marcelofg55 I had a look at the patch, looking at the diff it seems going in the right direction, sadly, it did not apply correctly on my branch, can you please make a PR so I can easily review it?

**Comment by akien-mga** [comment]
Fixed by bba8f1d.
